olive/auto_optimizer/__init__.py

```diff
@@ -0,0 +1,138 @@
+from olive.data.config import DataConfig
+from olive.evaluator.olive_evaluator import OliveEvaluatorConfig
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import ModelConfig
+
+logger = logging.getLogger(__name__)
+
+
+class Precision(str, Enum):
+    FP32 = "fp32"
+    FP16 = "fp16"
+    INT8 = "int8"
+    INT4 = "int4"
+
+
+class AutoOptimizerConfig(ConfigBase):
+    # opt_level
+    # 0: look up template mapping
+    # TODO(anyone): support more opt_level
+    opt_level: int = 0
+
+    # used to disable auto optimizer when user only want to evaluate input model
+    disable_auto_optimizer: bool = False
+
+    # precisions lists: [fp32, fp16, int8] which means the output model can be fp32, fp16 or int8
+    # which will be used to filter the passes that are not supported by the precision
+    # None for no precision restriction
+    precisions: Optional[List[Precision]] = None
+
+    # TODO(trajep): should distinguish model fine-tune and model inference?
+    # if fine_tune is True, we will not suggest the training related pass, like: QLora
+    # fine_tune: bool = False
+
+    @validator("opt_level", pre=True)
+    def check_opt_level(cls, v):
+        if v != 0:
+            raise ValueError("opt_level 0 is the only supported value for now")
+        if v not in [0, 1]:
+            raise ValueError("opt_level must be in [0, 1]")
+        return v
+
+
+class AutoOptimizer(RegulatePassConfigMixin):
+    def __init__(
+        self,
+        input_model_config: ModelConfig,
+        evaluator_config: OliveEvaluatorConfig,
+        accelerator_spec: AcceleratorSpec,
+        auto_optimizer_config: Optional[AutoOptimizerConfig] = None,
+        data_configs: Optional[Dict[str, DataConfig]] = None,
+    ):
+        # TODO(trajep): how to get model_attributes if the model is in AML workspace? Maybe we can ignore this for now?
+        self.input_model_config = input_model_config
+        self.evaluator_config = evaluator_config
+        self.accelerator_spec = accelerator_spec
+        self.auto_optimizer_config = auto_optimizer_config or AutoOptimizerConfig()
+        self.data_configs = data_configs or {}
+        self.initialize()
+
+    def initialize(self):
+        # 1. input model config
+        # TODO(trajep): move duplicate part of _prepare_non_local_model somewhere, maybe OliveSystem?
+        model_config = deepcopy(self.input_model_config)
+        resource_paths = model_config.get_resource_paths()
+        for resource_name, resource_path in resource_paths.items():
+            if not resource_path or resource_path.is_local_resource_or_string_name():
+                continue
+            downloaded_resource_path = cache_utils.download_resource(resource_path, self._config.cache_dir)
+            if downloaded_resource_path:
+                # set local resource path
+                model_config.config[resource_name] = downloaded_resource_path
+        model = model_config.create_model()
```

Comment:
>model [](http://example.com/codeflow?start=8&length=5)

where is it used?

Reply:
Removed and only keep input_model_config to regulate some passes roughly. Like for onnx model, we can skip conversion pass.